File: app/controllers/background_controller.rb
class BackgroundController < UITableViewController
  extend IB

  def viewDidLoad

    super

    @engine = Engine.sharedClient

    # fonts for use throughout
    @titleFont = UIFont.fontWithName(@engine.fontName, size: 18)

    @selectedPlaylist = @engine.ipod.selectedPlaylist
    @playlistImage = UIImage.imageNamed("playlist")
    @smartPlaylistImage = UIImage.imageNamed("smartplaylist")

  end

  def viewWillAppear(animated)

    super

    if @engine.ipod.playlist
      indexPath = NSIndexPath.indexPathForRow(@selectedPlaylist, inSection:0)         
      self.tableView.scrollToRowAtIndexPath(indexPath, atScrollPosition:UITableViewScrollPositionNone, animated:false)
    end

  end

  def dismiss

    self.dismissViewControllerAnimated(true, completion:nil)

  end


  #### Table View Delegate Methods ####

  def tableView(tableView, numberOfRowsInSection:section)

    @engine.ipod.allPlaylists.size

  end

  def tableView(tableView, titleForHeaderInSection:section)

    "Playlists"

  end

  def tableView(tableView, cellForRowAtIndexPath:indexPath)

    @reuseIdentifier ||= "CELL_IDENTIFIER"

    cell = tableView.dequeueReusableCellWithIdentifier(@reuseIdentifier)
    cell ||= UITableViewCell.alloc.initWithStyle(UITableViewCellStyleSubtitle, reuseIdentifier: @reuseIdentifier)
    
    # format cell
    cell.textLabel.font = @titleFont
    cell.detailTextLabel.font = @subtitleFont

    playlist = @engine.ipod.allPlaylists[indexPath.row]
    cell.textLabel.text = playlist.valueForProperty(MPMediaPlaylistPropertyName)

    if playlist.valueForProperty(MPMediaPlaylistPropertyPlaylistAttributes) == MPMediaPlaylistAttributeSmart
      cell.imageView.image = @smartPlaylistImage
    else
      cell.imageView.image = @playlistImage
    end


    if playlist == @engine.ipod.playlist
      cell.accessoryType = UITableViewCellAccessoryCheckmark
    else
      cell.accessoryType = UITableViewCellAccessoryNone
    end

    cell

  end

  def tableView(tableView, didSelectRowAtIndexPath:indexPath)

    # don't update playlist when it's already the current playlist
    unless indexPath.row == @selectedPlaylist
      @engine.ipod.usePlaylist(indexPath.row)
      self.presentingViewController.playlistDidChange

      # update checkmark and clear selection
      indexPaths = [NSIndexPath.indexPathForRow(@selectedPlaylist, inSection:0), indexPath]
      tableView.reloadRowsAtIndexPaths(indexPaths, withRowAnimation:UITableViewRowAnimationAutomatic)

      @selectedPlaylist = @engine.ipod.selectedPlaylist
    end

    tableView.deselectRowAtIndexPath(indexPath, animated:true)

  end

end

File: Stingtk/BackgroundController.swift
import UIKit
import MediaPlayer

class BackgroundController: UITableViewController {
    
    private let engine = Engine.sharedClient
    private var currentPlaylist: MPMediaPlaylist? {
        return engine.ipod.playlist
    }
    
    private var playlistImage = UIImage(named: "playlist")
    private var smartPlaylistImage = UIImage(named: "smartplaylist")
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if let playlistIndex = engine.ipod.playlistIndex {
            let indexPath = IndexPath(row: playlistIndex, section: 0)
            tableView.scrollToRow(at: indexPath, at: .none, animated: false)
        }
    }
    
    @IBAction func dismiss() {
        self.dismiss(animated: true, completion: nil)
    }
    
    // MARK: UITableViewDataSource
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return engine.ipod.allPlaylists.count
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "Playlists"
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Playlist Cell") ?? UITableViewCell(style: .default, reuseIdentifier: "Playlist Cell")
        
        let playlist = engine.ipod.allPlaylists[indexPath.row]
        cell.textLabel?.text = playlist.value(forProperty: MPMediaPlaylistPropertyName) as? String
        
        if let attributes = playlist.value(forProperty: MPMediaPlaylistPropertyPlaylistAttributes) as? MPMediaPlaylistAttribute {
            if attributes == .smart {
                cell.imageView?.image = smartPlaylistImage
            } else {
                cell.imageView?.image = playlistImage
            }
        }
        
        if playlist == currentPlaylist {
            cell.accessoryType = .checkmark
        } else {
            cell.accessoryType = .none
        }
        
        return cell
    }
    
    
    // MARK: UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let selectedPlaylist = engine.ipod.allPlaylists[indexPath.row]
        
        // don't update playlist when it's already the current playlist
        if selectedPlaylist != currentPlaylist {
            // TODO: Test ?? 0 for cases where selected is also 0
            let oldIndexPath = IndexPath(row: engine.ipod.playlistIndex ?? 0, section: 0)
            
            engine.ipod.usePlaylist(indexPath.row)
            (presentingViewController as! StkController).playlistDidChange()
            
            // update checkmark and clear selection
            let indexPaths = [oldIndexPath, indexPath]
            tableView.reloadRows(at: indexPaths, with: .automatic)
            
            save(engine.ipod)
        }
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    func save(_ ipod: Music) {
        // this needs to be changed to save by playlist id or similar
        // would enable checking of playlist properly if order changed
        UserDefaults.standard.set(ipod.playlistIndex, forKey: "Playlist Index")
    }
    
}
